File: python/c13_2_dijkstras_algorithm.py
import unittest
from heapq import heappush,heappop

INF=1<<63

def dijkstra(adj,x):
    distance=[INF]*len(adj)
    distance[x]=0
    q=[(0,x)]
    while q:
        m,a=heappop(q)
        if distance[a]<m:
            continue
        for b,w in adj[a]:
            if distance[a]+w<distance[b]:
                distance[b]=distance[a]+w
                heappush(q,(distance[b],b))
    return distance

class Test(unittest.TestCase):
    def test(self):
        self.assertEqual(dijkstra([[],[]],1),[INF,0])
        self.assertEqual(dijkstra([[],[(2,5),(4,9),(5,1)],[(1,5),(3,2)],[(2,2),(4,6)],[(1,9),(3,6),(5,2)],[(1,1),(4,2)]],1),[INF,0,5,7,3,1])

if __name__=='__main__':
    unittest.main()

File: c++/c13_2_dijkstras_algorithm.cpp
#include <bits/stdc++.h>
#include <gtest/gtest.h>
using namespace std;
using ll = long long;
constexpr ll INF = numeric_limits<ll>::max() / 2;

namespace {
	vector<ll> dijkstra(const vector<vector<pair<ll,ll>>>& adj, ll x) {
		ll n = ssize(adj);
		vector<ll> distance(n, INF);
		priority_queue<pair<ll, ll>, vector<pair<ll, ll>>, greater<>> q({}, { {0,x} });
		distance[x] = 0;
		while (!q.empty()) {
			auto [d, a] = q.top(); q.pop();
			if (distance[a] < d) continue;
			for (auto [b, w] : adj[a]) {
				if (distance[a] + w < distance[b]) {
					distance[b] = distance[a] + w;
					q.emplace(distance[b], b);
				}
			}
		}
		return distance;
	}
}

TEST(C132DijkstrasAlgorithm, dijkstra) {
	EXPECT_EQ(dijkstra({ {},{} }, 1), (vector<ll>{ INF,0 }));
	EXPECT_EQ(dijkstra({ {},{{2,5},{4,9},{5,1}},{{1,5},{3,2}},{{2,2},{4,6}},{{1,9},{3,6},{5,2}},{{1,1},{4,2}} }, 1), (vector<ll>{ INF,0,5,7,3,1 }));
}
